File: backend/app/core/permissions.py
"""Catálogo central de permissões e o conjunto padrão atribuído a cada role.

Esta é a única fonte de verdade sobre quais permissões existem no sistema.
Usada pelo seed (para popular `permissions` e `role_permissions`) e pelos
testes. Autorização em runtime nunca consulta este arquivo diretamente —
ela consulta as permissões efetivas do usuário no banco (role + overrides).
"""

MODULES = (
    "clients",
    "appointments",
    "services",
    "employees",
    "inventory",
    "finance",
    "reports",
    "settings",
    "audit",
    "notifications",
)

_ACTIONS_BY_MODULE: dict[str, tuple[str, ...]] = {
    "clients": ("view", "create", "edit", "delete"),
    "appointments": ("view", "create", "edit", "cancel", "confirm", "start", "complete"),
    "services": ("view", "create", "edit", "delete"),
    "employees": ("view", "create", "edit", "delete"),
    "inventory": ("view", "create", "edit", "adjust"),
    "finance": ("view", "create", "edit", "delete"),
    "reports": ("view",),
    "settings": ("view", "edit"),
    "audit": ("view",),
    "notifications": ("view", "manage"),
}

ALL_PERMISSIONS: list[str] = [
    f"{module}.{action}" for module, actions in _ACTIONS_BY_MODULE.items() for action in actions
]

ROLE_OWNER = "OWNER"
ROLE_MANAGER = "MANAGER"
ROLE_BARBER = "BARBER"
ROLE_RECEPTIONIST = "RECEPTIONIST"

ROLE_DEFINITIONS: dict[str, str] = {
    ROLE_OWNER: "Proprietário",
    ROLE_MANAGER: "Gerente",
    ROLE_BARBER: "Barbeiro",
    ROLE_RECEPTIONIST: "Recepcionista",
}

DEFAULT_ROLE_PERMISSIONS: dict[str, list[str]] = {
    # OWNER possui todas as permissões (seção 4).
    ROLE_OWNER: list(ALL_PERMISSIONS),
    ROLE_MANAGER: [
        "clients.view",
        "clients.create",
        "clients.edit",
        "appointments.view",
        "appointments.create",
        "appointments.edit",
        "appointments.cancel",
        "appointments.confirm",
        "appointments.start",
        "appointments.complete",
        "services.view",
        "services.create",
        "services.edit",
        "inventory.view",
        "inventory.create",
        "inventory.edit",
        "inventory.adjust",
        "finance.view",
        "employees.view",
        "reports.view",
        "notifications.view",
    ],
    ROLE_BARBER: [
        "appointments.view",
        "appointments.confirm",
        "appointments.start",
        "appointments.complete",
        "clients.view",
        "services.view",
        "notifications.view",
    ],
    ROLE_RECEPTIONIST: [
        "clients.view",
        "clients.create",
        "clients.edit",
        "appointments.view",
        "appointments.create",
        "appointments.edit",
        "appointments.cancel",
        "appointments.confirm",
        "services.view",
        "finance.view",
        "finance.create",
        "notifications.view",
    ],
}

File: frontend/src/lib/permissions.ts
/**
 * Catálogo de cargos e permissões — espelha `backend/app/core/permissions.py`
 * (fonte de verdade). Não existe endpoint que devolva esse catálogo, então
 * replicamos aqui, igual a outros enums estáveis do projeto (status, dias
 * da semana etc.).
 */
export const STAFF_ROLES = [
  { code: "OWNER", name: "Proprietário" },
  { code: "MANAGER", name: "Gerente" },
  { code: "BARBER", name: "Barbeiro" },
  { code: "RECEPTIONIST", name: "Recepcionista" },
] as const;

const ACTION_LABELS: Record<string, string> = {
  view: "Ver",
  create: "Criar",
  edit: "Editar",
  delete: "Excluir",
  cancel: "Cancelar",
  confirm: "Confirmar",
  start: "Iniciar",
  complete: "Finalizar",
  adjust: "Ajustar",
  manage: "Gerenciar",
  reply: "Responder",
};

export const PERMISSION_MODULES: { module: string; label: string; actions: string[] }[] = [
  { module: "clients", label: "Clientes", actions: ["view", "create", "edit", "delete"] },
  {
    module: "appointments",
    label: "Agendamentos",
    actions: ["view", "create", "edit", "cancel", "confirm", "start", "complete"],
  },
  { module: "services", label: "Serviços", actions: ["view", "create", "edit", "delete"] },
  { module: "employees", label: "Profissionais e equipe", actions: ["view", "create", "edit", "delete"] },
  { module: "inventory", label: "Estoque", actions: ["view", "create", "edit", "adjust"] },
  { module: "finance", label: "Financeiro", actions: ["view", "create", "edit", "delete"] },
  { module: "reports", label: "Relatórios", actions: ["view"] },
  { module: "settings", label: "Configurações", actions: ["view", "edit"] },
  { module: "audit", label: "Auditoria", actions: ["view"] },
  { module: "notifications", label: "Notificações", actions: ["view", "manage"] },
  { module: "messages", label: "Mensagens", actions: ["view", "reply"] },
];

export function actionLabel(action: string): string {
  return ACTION_LABELS[action] ?? action;
}

export function roleName(code: string): string {
  return STAFF_ROLES.find((r) => r.code === code)?.name ?? code;
}
